include/verilated_replay.cpp

```diff
@@ -0,0 +1,223 @@
+//                return -1;
+//            }
+//
+//            if ((*it)->type != VCD_VAR_REG && (*it)->type != VCD_VAR_WIRE) {
+//                VL_PRINTF("Error unsupported signal type on %s\n", signalName.c_str());
+//                return -1;
+//            }
+//
+//            m_inputs.push_back(signalp);
+//            return 0;
+//        }
+//    }
+//    VL_PRINTF("Error finding signal (%s)\n", signalName.c_str());
+//    return -1;
+//}
+//
+//void VerilatedReplay::copySignal(uint8_t* signal, VCDValue* valuep) {
+//    VCDValueType type = valuep->get_type();
+//    switch(type) {
+//        case VCD_SCALAR: {
+//            copySignalBit(signal, 0, valuep->get_value_bit());
+//            break;
+//        }
+//        case VCD_VECTOR: {
+//            VCDBitVector* bitVector = valuep->get_value_vector();
+//            unsigned length = bitVector->size();
+//            for (int i = 0; i < length; ++i) {
+//                copySignalBit(signal, i, bitVector->at(length - i - 1));
+//            }
+//            break;
+//        }
+//        default: {
+//            VL_PRINTF("Error unsupported VCD value type");
+//            exit(-1);
+//        }
+//    }
+//}
+//
+//void VerilatedReplay::copySignalBit(uint8_t* signal, unsigned offset, VCDBit bit) {
+//    unsigned byte = offset / 8;
+//    unsigned byteOffset = offset % 8;
+//    // TODO - more efficient byte copying
+//    signal[byte] &= ~(0x1 << byteOffset);
+//    // TODO - x's and z's?
+//    signal[byte] |= (bit == VCD_1 ? 0x1 : 0x0) << byteOffset;
+//}
+
+int VerilatedReplay::replay() {
+    // TODO -- lockless ring buffer for separate reader/replay threads
+    // TODO -- should I be using fstReaderIterBlocks instead? (only one CB)
+    // It appears that 0 is the error return code
+    if (fstReaderIterBlocks2(m_fstp, &VerilatedReplay::fstCallback,
+                             &VerilatedReplay::fstCallbackVarlen, this, NULL) == 0) {
+        VL_PRINTF("Error iterating FST\n");
+        exit(-1);
+    }
+
+    // One final eval + trace since we only eval on time changes
+    eval();
+    trace();
+    final();
+
+    return 0;
+}
+
+void VerilatedReplay::fstCb(uint64_t time, fstHandle facidx,
+                            const unsigned char* valuep, uint32_t len) {
+    // Watch for new time steps and eval before we start working on the new time
+    if (m_time != time) {
+        eval();
+        trace();
+        m_time = time;
+        // TODO -- use FST timescale
+        m_simTime = m_time;
+    }
+
+    // TODO -- remove
+    VL_PRINTF("%lu %u %s\n", time, facidx, valuep);
+
+    // TODO -- is len always right, or should we use strlen() or something?
+    // TODO -- handle values other than 0/1, what can show up here?
+    vluint8_t* signal = m_inputHandles[facidx].signal;
+    vluint8_t byte = 0;
+    for (size_t bit = 0; bit < len; ++bit) {
+        char value = valuep[len - 1 - bit];
+        if (value == '1') byte |= 1 << (bit % 8);
```

Comment:
This will be slow for the same reason that writing FSTs is slow today.  Hopefully we can get to a binary API for both reading and writing FSTs someday.

In any case, even if we had a binary API here, I'm still going to not want to be slowed down by reading the file.  I anticipate pulling in the entire FST data set or pulling in enough via a reader thread so as to not have the sim slowed down by FST I/O.